### PR TITLE
Refactor help internals into separate interface/class

### PR DESCRIPTION
**Pull request by shadowspawn**
# Pull Request

Major help refactor (see also earlier experimentation in #1338, #1346). 

## Problem

The help generation is an organic collection of methods with custom logic embedded at various levels, and inconsistent handling of arguments, options, and commands. 

- the code is messy to maintain and extend
- it isn't easy for programs to modify or build on the built-in processing
    - nice example of complex help: https://github.com/tj/commander.js/pull/1338#issuecomment-683239306
- there isn't a natural place in API to add functionality, like sorting the help

Related issues:
- #1338 custom help
- #625 #910 sorting
- #301 #621 color

## Solution

Move the help related routines into a separate class. Inspired by discussion and examples from @cravler: https://github.com/tj/commander.js/pull/1346#issuecomment-689476637

Implement the Help methods as if they are static to allow override not only with subclass (see `.createHelp()`), but also with plain  functions (see `.configureHelp()`).

The impact on existing programs only using public methods should be small. All of the methods which are moved or deleted in the refactor were private, except for `Option.fullDescription` which is new in v7 so not released yet.

Callers can:
- subclass Help for complex cases, such as styling output with colour
- configure Help for simpler cases

For example:

```js
program.configureHelp({
   sortSubcommands: true,
   subcommandTerm: (cmd) => cmd.name()
});
```

## ChangeLog

- (added) refactor the code generating the help into a separate public Help class
    - support sorting subcommands and options in help
    - support specifying wrap width (columns)
    - allow subclassing Help class
    - allow configuring Help class without subclassing
- (changed) initialise the command description to empty string (previously undefined)
- (fixed) in the help the first line of command description was wrapping two characters early
- (fixed) in the help the pad width calculation was not including help option and help command
- (fixed) in the help the pad width calculation was including hidden options and commands

### CODE REVIEW

**Comment by shadowspawn** [comment]
Here is an example:

```js
const { program } = require('commander');

function formatHelpAsMarkdown(cmd, helper) {
  function formatItem(term, description) {
    if (description) {
      return `\`${term}\`: ${description}`;
    }
    return `\`${term}\``;
  };
  function formatList(textArray) {
    return textArray.join('\n').replace(/^/gm, '* ');
  }

  let output = [`# ${cmd.name()}`, ''];

  // Usage
  output = output.concat([helper.commandUsage(cmd), '']);

  // Description
  if (cmd.description()) {
    output = output.concat([cmd.description(), '']);
  }

  // Arguments
  const visibleArguments = helper.visibleArguments(cmd);
  if (visibleArguments.length) {
    const argumentsList = visibleArguments.map((argument) => {
      return formatItem(argument.term, argument.description);
    });
    output = output.concat(['## Arguments:', formatList(argumentsList), '']);
  }

  // Options
  const visibleOptions = helper.visibleOptions(cmd);
  if (visibleOptions.length) {
    const optionList = visibleOptions.map((option) => {
      return formatItem(helper.optionTerm(option), helper.optionDescription(option));
    });
    output = output.concat(['## Options:', formatList(optionList), '']);
  }

  // Commands
  const visibleCommands = helper.visibleCommands(cmd);
  if (visibleCommands.length) {
    const commandList = visibleCommands.map((cmd) => {
      return formatItem(helper.commandTerm(cmd), helper.commandDescription(cmd));
    });
    output = output.concat(['## Commands:', formatList(commandList), '']);
  }

  return output.join('\n');
}

function commandUsage(cmd) {
  return `Usage: \`${cmd.name()} ${cmd.usage()}\``;
}

program
  .name('mighty-markdown')
  .helpOverrides({ formatHelp: formatHelpAsMarkdown, commandUsage: commandUsage })
  .option('-f, --foo', 'first option')
  .option('-b, --bar', 'first option')
  .command('sub1', 'external')
  .command('sub2', 'external')


program.parse();
```

------

```sh
node markdown.sh --help
# paste in below
```

# mighty-markdown

Usage: `mighty-markdown [options] [command]`

## Options:
* `-f, --foo`: first option
* `-b, --bar`: first option
* `-h, --help`: display help for command

## Commands:
* `sub1`: external
* `sub2`: external
* `help [command]`: display help for command

**Comment by shadowspawn** [comment]
And a much smaller example, setting the wrapping width for say running a unit test:

```js
program.helpOverrides({ columns: 80 })
```

**Comment by cravler** [comment]
Hi @shadowspawn, 

Played around with new `HelpTools` and it looks very flexible:

```js
const { Command, HelpTools } = require('./index');

class Help {
  constructor(cmd, helper) {
    this.cmd = cmd;
    this.helper = helper;
  };

  usage() {
    return this.helper.commandUsage(this.cmd);
  };

  description() {
    return this.helper.commandDescription(this.cmd);
  };

  arguments() {
    return this.helper.visibleArguments(this.cmd);
  };

  options() {
    return this.helper.visibleOptions(this.cmd).map(option => ({
      term: this.helper.optionTerm(option),
      description: this.helper.optionDescription(option),
    }));
  };

  commands() {
    return this.helper.visibleCommands(this.cmd).map(cmd => ({
      term: this.helper.commandTerm(cmd),
      description: this.helper.commandDescription(cmd),
    }));
  };

  render() {
    const params = this.helper.formatParams();

    const baseIndentWidth = params.baseIndentWidth || 0;
    const itemIndentWidth = params.itemIndentWidth || 2;
    const itemSeparatorWidth = params.itemSeparatorWidth || 2;

    const columns = this.helper.columns;
    const termWidth = this.helper.padWidth(this.cmd, this.helper) + baseIndentWidth;
    const descriptionWidth = columns - termWidth - baseIndentWidth - itemIndentWidth - itemSeparatorWidth;

    const styles = this.helper.styles();

    const formatList = (arr) => {
      return arr.map(({ term, description }) => {
        if (description) {
          return [
            styles.term(this.helper.pad(term, termWidth + itemSeparatorWidth)),
            this.helper.optionalWrap(description, descriptionWidth, termWidth + itemSeparatorWidth, this.helper),
          ].join('');
        }
        return term;
      }).join('\n').replace(/^/gm, Array(itemIndentWidth + 1).join(' '));
    };

    const labels = this.helper.labels();
    Object.keys(labels).map(key => {
      labels[key] = styles.label(labels[key]);
    });

    // Usage
    let output = [
      labels.usage,
      Array(itemIndentWidth + 1).join(' ') + this.usage(),
      '',
    ];

    // Description
    if (this.description()) {
      output = output.concat([
        Array(itemIndentWidth + 1).join(' ') + styles.description(this.helper.optionalWrap(
          this.description(),
          columns - baseIndentWidth - itemIndentWidth,
          itemIndentWidth,
          this.helper
        )),
        '',
      ]);
    }

    // Arguments
    if (this.arguments().length) {
      output = output.concat([
        labels.arguments,
        formatList(this.arguments()),
        '',
      ]);
    }

    // Options
    if (this.options().length) {
      output = output.concat([
        labels.options,
        formatList(this.options()),
        '',
      ]);
    }

    // Commands
    if (this.commands().length) {
      output = output.concat([
        labels.commands,
        formatList(this.commands()),
        '',
      ]);
    }

    return output.join('\n').replace(/^/gm, Array(baseIndentWidth + 1).join(' ')).trimEnd() + '\n';
  };
}

class MyHelpTools extends HelpTools {
  labels() {
    return {
      usage: 'Usage:',
      arguments: 'Arguments:',
      options: 'Options:',
      commands: 'Commands:',
    };
  };

  styles() {
    return {
      label: str => str,
      term: str => str,
      description: str => str,
    };
  };

  commandUsage(cmd) {
    return super.commandUsage(cmd).split('Usage: ')[1];
  };

  formatParams() {
    return {
      baseIndentWidth: 0,
      itemIndentWidth: 2,
      itemSeparatorWidth: 2
    };
  };

  formatHelp(cmd, helper) {
    return ['', (new Help(cmd, helper)).render(), ''].join('\n');
  };
}

class MyCommand extends Command {
  createCommand(name) {
    return new MyCommand(name);
  };

  createHelpTools() {
    return Object.assign(new MyHelpTools(), this._helpToolsOverrides);
  };
}

const program = new MyCommand('example');

program.version('0.0.1');

program.helpToolsOverrides({
  labels: () => {
    return {
      usage: '>>> Usage:\n',
      arguments: '>>> Arguments:\n',
      options: '>>> Options:\n',
      commands: '>>> Commands:\n',
    }
  },
  styles: () => {
    return {
      label: str => '\x1b[33m' + str + '\x1b[0m', // yellow
      term: str => '\x1b[32m' + str + '\x1b[0m', // green
      description: str => '\x1b[36m' + str + '\x1b[0m', // cyan
    };
  },
  formatParams () {
    return {
      baseIndentWidth: 2,
    };
  },
});

program
  .command('test <cmd> [env] [val]')
  .description(Array(100).join('long description '), {
    cmd: 'cmd description',
    env: 'env description',
    val: Array(50).join('val description ')
  })
  .action((cmd, env, options) => {});

program.parse(process.argv);
```

**Comment by shadowspawn** [comment]
Thanks @cravler . I was wondering about applying styles, and that is just one of the layers you have added, and even used both subclass and arrow function customisations. Thanks for the big try-out!

(I changed `HelpTools` to `Help` in the PR today, apologies for the churn.)

**Comment by cravler** [comment]
Hi @shadowspawn,

It wold be good to change `cmd.helpOverrides()` to:

```js
class Command {
  helpOverrides(overrides) {
    if (overrides === undefined) return this._helpOverrides;
    this._helpOverrides = overrides;
    return this;
  };
}
```
Then we do not need to access priate `cmd._helpOverrides`:

```js
class MyCommand extends Command {
  createHelp() {
    return Object.assign(new MyHelp(), this.helpOverrides());
  };
}
```

**Comment by shadowspawn** [comment]
Musing about colour. Related issues: #301 #621 #779

This PR does make it a bit easier to add big help customisations, including colour as in https://github.com/tj/commander.js/pull/1365#issuecomment-701265028

Adding colour is fairly straightforward if applied to a whole logical piece of text. It is trickier when applied to parts of text, like in the picture in #301. Setting the colour on the various pieces of text is intrusive in the text construction, and the pad calculations should not include the styling characters. Or could be offered by lots of style hooks, like https://sywac.io/docs/style-hooks.html. Or templates.

I am planning to leave direct support for colour/styling out of this PR.

**Comment by shadowspawn** [comment]
Removed own `pad()` routine and replaced with `string.padEnd()`.

**Comment by shadowspawn** [comment]
Renamed `helpOverrides()` to `.configureHelp()`.

**Comment by shadowspawn** [comment]
Rolled together `optionalWrap()` and `wrap()`, now just `wrap()`.

**Comment by shadowspawn** [comment]
`.commandDescription()` is ambiguous out of context of the term/description pattern for the arguments, options, and (sub)commands. It get uses for the subcommand descriptions and not for the main command. I haven't come up with clear naming that I like better, so my current plan is be clear in the JSDoc/TSDoc!

**Comment by cravler** [comment]
Example how to use `short` & `long` description:

```js
const { program } = require('./index');

class CommandDescription {
  constructor({ short = '', long = '' } = {}) {
    this.short = short;
    this.long = long;
  };

  toString() {
    return this.long;
  };
}

program.configureHelp({
  commandDescription: cmd => {
    const description = cmd.description();
    if (description instanceof CommandDescription) {
      return description.short;
    }
    return description;
  }
});

program
  .command('test <cmd> [env]')
  .description(new CommandDescription({
    short: 'short description',
    long: Array(100).join('long description ')
  }), {
    cmd: 'cmd description',
    env: 'env description'
  })
  .action((cmd, env, options) => {});

program.parse(process.argv);
```

File: index.js [hunk 1]
```diff
@@ -9,13 +9,328 @@ const fs = require('fs');
...
+    };
+    function formatList(textArray) {
+      return textArray.join('\n').replace(/^/gm, ' '.repeat(itemIndentWidth));
+    }
+
+    // Usage
+    let output = [helper.commandUsage(cmd), ''];
+
+    // Description
+    if (cmd.description()) {
```
Comment (cravler) [inline index.js:249]: better to use: `cmd.description().length`

Comment (shadowspawn) [reply index.js:249]: Can do that now (until 3 days ago it was undefined when not set 😄 ).

Comment (shadowspawn) [reply index.js]: Expanded all the way to `.length > 0`.

File: index.js [hunk 2]
```diff
@@ -9,13 +9,340 @@ const fs = require('fs');
...
+    // Usage
+    let cmdName = cmd._name;
+    if (cmd._aliases[0]) {
+      cmdName = cmdName + '|' + cmd._aliases[0];
+    }
+    let parentCmdNames = '';
+    for (let parentCmd = cmd.parent; parentCmd; parentCmd = parentCmd.parent) {
+      parentCmdNames = parentCmd.name() + ' ' + parentCmdNames;
+    }
+    return 'Usage: ' + parentCmdNames + cmdName + ' ' + cmd.usage();
```
Comment (cravler) [inline index.js:179]: maybe better to return without `Usage:` label prefix?

Comment (shadowspawn) [reply index.js:179]: Yes, that would be more consistent. I'll do that, thanks.

**Comment by cravler** [comment]
Updated example how to use `short` & `long` description:

```js
const { program } = require('./index');

program.configureHelp({
  commandDescription: cmd => {
    const description = cmd.description();
    if (typeof description === 'object') {
      return description.long || '';
    }
    return description;
  },
  subcommandDescription: cmd => {
    const description = cmd.description();
    if (typeof description === 'object') {
      return description.short || '';
    }
    return description;
  }
});

program
  .command('test <cmd> [env]')
  .description({
    short: 'short description',
    long: Array(100).join('long description ')
  }, {
    cmd: 'cmd description',
    env: 'env description'
  })
  .action((cmd, env, options) => {});

program.parse(process.argv);

```

**Comment by shadowspawn** [comment]
> Updated example how to use short & long description:

This is clever, but I'll note it is not supported and may break in the future. The interface for `.description()` lists a "string" for the first parameter. The internal variable is hardly touched though, so the override works fine at the moment!

**Comment by shadowspawn** [comment]
Ready for review. This is a big refactor of existing code, and makes the methods used for building the help public so they may be configured in custom help implementations.

File: index.js [hunk 3]
```diff
@@ -9,13 +9,344 @@ const fs = require('fs');
...
+        // use stringify to match the display of the default value
+        `choices: ${option.argChoices.map((choice) => JSON.stringify(choice)).join(', ')}`);
+    }
+    if (option.defaultValue !== undefined) {
+      extraInfo.push(`default: ${option.defaultValueDescription || JSON.stringify(option.defaultValue)}`);
+    }
+    if (extraInfo.length > 0) {
+      return `${option.description} (${extraInfo.join(', ')})`;
+    }
+    return `${option.description}`;
```
Comment (abetomo) [inline index.js:233]: '`' seems to be unnecessary.

Comment (shadowspawn) [reply index.js]: Yes, removed.

File: index.js [hunk 4]
```diff
@@ -9,13 +9,344 @@ const fs = require('fs');
...
+
+  /**
+   * Get the option term to show in the list of options.
+   *
+   * @param {Option} option
+   * @returns {string}
+   */
+
+  optionTerm(option) {
+    return `${option.flags}`;
```
Comment (abetomo) [inline index.js:121]: It seems to be OK to return the variable as it is.

Comment (shadowspawn) [reply index.js]: Another one! I think I had some problems with one of the TypeScript tests during development, but no longer the case. Removed the template literal.

**Comment by shadowspawn** [comment]
(I'll hopefully have time over next week to prepare a prerelease of 7.x, and a 6.x as well.)

**Review by abetomo — APPROVED** [review]
LGTM.
That is great! Thank you!

**Comment by shadowspawn** [comment]
Included in v7.0.0 pre-release #1386

**Pull request merged**